vng/vng_test.go

```diff
@@ -0,0 +1,351 @@
+		valueOut := valuesOut[i]
+		t.Logf("comparing: in[%v]=%v vs out[%v]=%v", i, zson.String(&valueIn), i, zson.String(&valueOut))
+		if !bytes.Equal(zed.EncodeTypeValue(valueIn.Type), zed.EncodeTypeValue(valueOut.Type)) {
+			t.Errorf("values have different types: %v vs %v", valueIn.Type, valueOut.Type)
+		}
+		if !bytes.Equal(valueIn.Bytes(), valueOut.Bytes()) {
+			t.Errorf("values have different zng bytes: %v vs %v", valueIn.Bytes(), valueOut.Bytes())
+		}
+	}
+	for i := range valuesOut[len(valuesIn):] {
+		t.Errorf("extra value: out[%v].Bytes()=%v", i, valuesOut[i].Bytes())
+		t.Errorf("extra value: out[%v]=%v", i, zson.String(&valuesOut[i]))
+	}
+}
+
+func genValues(b *bytes.Reader, context *zed.Context, types []zed.Type) []zed.Value {
+	values := make([]zed.Value, 0)
+	var builder zcode.Builder
+	for genByte(b) != 0 {
+		typ := types[int(genByte(b))%len(types)]
+		builder.Reset()
+		genValue(b, context, typ, &builder)
+		it := builder.Bytes().Iter()
+		values = append(values, *zed.NewValue(typ, it.Next()).Copy())
+	}
+	return values
+}
+
+func genValue(b *bytes.Reader, context *zed.Context, typ zed.Type, builder *zcode.Builder) {
+	if genByte(b) == 0 {
+		builder.Append(nil)
+		return
+	}
+	switch typ {
+	case zed.TypeUint8:
+		builder.Append(zed.EncodeUint(uint64(genByte(b))))
+	case zed.TypeUint16:
+		builder.Append(zed.EncodeUint(uint64(binary.LittleEndian.Uint16(genBytes(b, 2)))))
+	case zed.TypeUint32:
+		builder.Append(zed.EncodeUint(uint64(binary.LittleEndian.Uint32(genBytes(b, 4)))))
+	case zed.TypeUint64:
+		builder.Append(zed.EncodeUint(uint64(binary.LittleEndian.Uint64(genBytes(b, 8)))))
+	case zed.TypeInt8:
+		builder.Append(zed.EncodeInt(int64(genByte(b))))
+	case zed.TypeInt16:
+		builder.Append(zed.EncodeInt(int64(binary.LittleEndian.Uint16(genBytes(b, 2)))))
+	case zed.TypeInt32:
+		builder.Append(zed.EncodeInt(int64(binary.LittleEndian.Uint32(genBytes(b, 4)))))
+	case zed.TypeInt64:
+		builder.Append(zed.EncodeInt(int64(binary.LittleEndian.Uint64(genBytes(b, 8)))))
+	case zed.TypeDuration:
+		builder.Append(zed.EncodeDuration(nano.Duration(int64(binary.LittleEndian.Uint64(genBytes(b, 8))))))
+	case zed.TypeTime:
+		builder.Append(zed.EncodeTime(nano.Ts(int64(binary.LittleEndian.Uint64(genBytes(b, 8))))))
+	case zed.TypeFloat16:
+		panic("Unreachable")
+	case zed.TypeFloat32:
+		builder.Append(zed.EncodeFloat32(math.Float32frombits(binary.LittleEndian.Uint32(genBytes(b, 4)))))
+	case zed.TypeFloat64:
+		builder.Append(zed.EncodeFloat64(math.Float64frombits(binary.LittleEndian.Uint64(genBytes(b, 8)))))
+	case zed.TypeBool:
+		builder.Append(zed.EncodeBool(genByte(b) > 0))
+	case zed.TypeBytes:
+		builder.Append(zed.EncodeBytes(genBytes(b, int(genByte(b)))))
+	case zed.TypeString:
+		builder.Append(zed.EncodeString(string(genBytes(b, int(genByte(b))))))
+	case zed.TypeIP:
+		builder.Append(zed.EncodeIP(netip.AddrFrom16(*(*[16]byte)(genBytes(b, 16)))))
+	case zed.TypeNet:
+		ip := netip.AddrFrom16(*(*[16]byte)(genBytes(b, 16)))
```

Comment:
No pointers required since we're using Go 1.20.
```suggestion
		builder.Append(zed.EncodeIP(netip.AddrFrom16([16]byte)(genBytes(b, 16))))
	case zed.TypeNet:
		ip := netip.AddrFrom16([16]byte)(genBytes(b, 16))
```